lib/puppet/network/http/connection.rb

```diff
@@ -205,7 +206,14 @@ def apply_options_to(request, options)
     end
 
     def execute_request(connection, request)
+      start = Time.now
       connection.request(request)
+    rescue EOFError
+      elapsed = (Time.now - start).to_f.round(3)
+      uri = @site.addr + request.path.split('?')[0]
+      eof = EOFError.new(_('request %{uri} interrupted after %{elapsed} seconds') % {uri: uri, elapsed: elapsed})
+      eof.set_backtrace(caller)
```

Comment:
One problem with this is that it will set the backtrace to the location of the `rescue` block, which looses information about where the error was originally raised.

Instead of using `caller`, we can copy the backtrace from the original exception:

```
rescue EOFError => e
  elapsed = (Time.now - start).to_f.round(3)
  uri = @site.addr + request.path.split('?')[0]
  eof = EOFError.new(_('request %{uri} interrupted after %{elapsed} seconds') % {uri: uri, elapsed: elapsed})
  eof.set_backtrace(e.backtrace) # Copy backtrace from rescued exception
  raise eof
end

Reply:
@Sharpie frustratingly, the original exception didn't come with a backtrace in my reproduction. That made it really annoying to track down this far. My assumption then was that such an EOF would never have a backtrace. If you think that's not true, we could do some logic to conditionally set the backtrace if it's empty, and preserve if it exists.

Do you know why this EOFError didn't have a backtrace?

Reply:
Oh wow. Yeah, if the original EOFError didn't have a backtrace, then this is definitely better than nothing.

The only thing I can think of off the top of my head that would cause that is if the EOFError is created in Ruby C code that doesn't bother to attach a Ruby-level stacktrace.

Reply:
@joshcooper updated and new commit pushed

Reply:
@joshcooper actually wait, that code doesn't work as-is... working on it